modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ *
+ * typealias Job<A> = Reader<Context, A>
+ *
+ * fun processTopic(t: Topic): Job<Result> = TODO()
+ * ```
+ *
+ * We can imagine we have a data pipeline that processes a bunch of data, each piece of data being categorized by a topic. Given a specific topic, we produce a `Job` that processes that topic. (Note that since a `Job` is just a `Reader`/`Kleisli`, one could write many small `Jobs` and compose them together into one `Job` that is used/returned by `processTopic`.)
+ *
+ * Corresponding to bunch of topics, a `List<Topic>` if you will. Since `Reader` has an [Applicative] instance, we can [traverse] over this list with `processTopic`.
+ *
+ * ```kotlin:ank
+ * import arrow.core.Id
+ * import arrow.core.ForId
+ * import arrow.core.ListK
+ * import arrow.core.extensions.id.applicative.applicative
+ * import arrow.mtl.KleisliPartialOf
+ * import arrow.mtl.extensions.kleisli.applicative.applicative
+ * import arrow.mtl.fix
+ *
+ * val JobForContext = Job.applicative<ForId, Context>(Id.applicative())
+ *
+ * fun processTopics(topics: ListK<Topic>): Job<ListK<Result>> =
+ *   topics.traverse<KleisliPartialOf<ForId, Context>, Result>(JobForContext) {
+ *     processTopic(it)
+ *   }.fix()
+ * ```
+ *
+ * Note the nice return type - `Job<List<Result>>`. We now have one aggregate `Job` that when run, will go through each topic and run the topic-specific job, collecting results as it goes. We say "when run" because a `Job` is some function that requires a Context before producing the value we want.
+ *
+ * One example of a "context" can be found in the [Spark](http://spark.apache.org/) project or in [Android's Context](https://developer.android.com/reference/android/content/Context). In Spark, information needed to run a Spark job (where the master node is, memory allocated, etc.) resides in a `SparkContext`. Going back to the above example, we can see how one may define topic-specific Spark jobs `(type Job<A> = Reader<SparkContext, A>)` and then run several Spark jobs on a collection of topics via [traverse]. We then get back a `Job<List<Result>>`, which is equivalent to (`SparkContext) -> List<Result>`. When finally passed a `SparkContext`, we can run the job and get our results back.
+ *
+ * Moreover, the fact that our aggregate job is not tied to any specific `SparkContext` allows us to pass in a `SparkContext` pointing to a production cluster, or (using the exact same job) pass in a test `SparkContext` that just runs locally across threads. This makes testing our large job nice and easy.
+ *
+ * Finally, this encoding ensures that all the jobs for each topic run on the exact same cluster. At no point do we manually pass in or thread a `SparkContext` through - that is taken care for us by the (applicative) behavior of `Reader` and therefore by [traverse].
+ *
+ * ### Traversables are Functors
+ *
+ * As it turns out every [Traverse] is a lawful [Functor]. By carefully picking the `G` to use in [traverse] we can implement `map`.
```

Comment:
This is a duplicate, but I like this section much more ^^

Reply:
The intention here is to further highlight the power and versatility of `Traverse`, through different angles.